### PR TITLE
Initial review

### PR DESCRIPTION
**Pull request by LogvinovLeon**
This PR:
* Is a fake PR to give others an ability to review this code and leave the comments
* It implements initial mempool aggregation and processing
* Also adds simple test

### CODE REVIEW

File: migrations/001-initial.sql [hunk 1]
```diff
@@ -0,0 +1,30 @@
+-- Up
```
Comment (fabioberger) [inline migrations/001-initial.sql]: Can we stay with `underscore` file naming convention please?

File: migrations/001-initial.sql [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+CREATE TABLE tx (
+    hash text PRIMARY KEY,
+    block_hash text,
+    FOREIGN KEY (block_hash) REFERENCES block(hash)
+);
+
+CREATE TABLE mempool (
+    timestamp INTEGER,
+    tx_hash text,
+    node_url text,
```
Comment (fabioberger) [inline migrations/001-initial.sql]: So are we sure the nodeUrl won't ever change for the life-time of the node?

File: package.json [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    "@types/bignumber.js": "^4.0.3",
+    "@types/body-parser": "^1.16.5",
+    "@types/express": "^4.0.37",
+    "@types/lodash": "^4.14.76",
+    "@types/mocha": "^2.2.43",
+    "@types/mockery": "^1.4.29",
+    "@types/node": "^8.0.31",
+    "@types/node-fetch": "^1.6.7",
+    "@types/sinon": "^2.3.5",
+    "bignumber.js": "^4.1.0",
```
Comment (fabioberger) [inline package.json]: Shouldn't this be a dep?

File: src/db.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+export class DB {
+    private db: Database;
+    constructor(db: Database) {
+        this.db = db;
+    }
+    public async getTransactions(): Promise<any> {
+        const transactions = await this.db.all('SELECT * FROM tx');
+        return transactions;
+    }
+    public async putNodeIfNotExists(url: string, region: string, version: string): Promise<void> {
```
Comment (fabioberger) [inline src/db.ts]: rename to `addNodeIfNotExistsAsync`

File: src/db.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        await this.db.run('INSERT OR IGNORE INTO node(url, region, version) VALUES(?, ?, ?)', [url, region, version]);
+    }
+    public async putMempoolTxIfNotExists(tx_hash: string, node_url: string): Promise<void> {
```
Comment (fabioberger) [inline src/db.ts]: `addMempoolTxIfNotExistsAsync`

Comment (fabioberger) [reply src/db.ts]: Don't use underscores in variable names except in Sqlite. Use Camel-case instead.

File: src/error_reporter.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
+import * as express from 'express';
+import {utils} from './utils';
+import {configs} from './configs';
+import rollbar = require('rollbar');
```
Comment (fabioberger) [inline src/error_reporter.ts]: Put external exports to top.

File: src/index.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import * as bodyParser from 'body-parser';
+import * as express from 'express';
+import wrap = require('express-async-wrap');
```
Comment (fabioberger) [inline src/index.ts]: need to use `require` here?

File: src/index.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+app.use(bodyParser.json()); // for parsing application/json
+app.use((req, res, next) => {
+  res.header('Access-Control-Allow-Origin', '*');
+  res.header('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept');
+  next();
+});
+
+const main = (async () => {
+    const database = await sqlite.open('./database.sqlite');
+    await database.migrate({ force: 'last' });
```
Comment (fabioberger) [inline src/index.ts]: remove spaces around brackets.

File: src/index.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    const db = new DB(database);
```
Comment (fabioberger) [inline src/index.ts]: My vote would be to do:

```
const database = await sqlite.open('./database.sqlite');
await database.migrate({force: 'last'});
```

inside the `DB` class constructor. I'm guessing this class is a singleton?

File: src/index.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    const mempool = new Mempool(db, {
+        'http://13.114.186.220:8545': 'Tokyo',
+        'http://13.114.106.120:8545': 'Tokyo',
+        'http://54.215.190.250:8545': 'SFO',
+        'http://54.67.64.189:8545': 'SFO',
+        'http://18.194.40.4:8545': 'FRA',
+        'http://52.58.239.93:8545': 'FRA',
```
Comment (fabioberger) [inline src/index.ts]: Put these in a `configs` module and the city codes in an enum.

Comment (fabioberger) [reply src/index.ts]: Let's not pass this as an anonymous object.

File: src/index.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    });
+    await mempool.saveNodeInfoAsync();
```
Comment (fabioberger) [inline src/index.ts]: Maybe put a comment above this that it's only here for development. Eventually should be removed. I'm guessing the DB entries of nodes is  going to be long-lived? Adding a node shouldn't require a service re-start right? Can simply add it to the DB from the sqlite console.

File: src/mempool.ts [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+    'gas',
+    'gasPrice',
+    'from',
+    'to',
+    'blockNumber',
+];
+
+export class Mempool {
+    private nodeRegionByUrl: {[nodeUrl: string]: string};
+    private nodeVersion: {[nodeUrl: string]: string};
```
Comment (fabioberger) [inline src/mempool.ts]: Rename to `nodeVersionByUrl`

File: src/mempool.ts [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+    private id = 0;
```
Comment (fabioberger) [inline src/mempool.ts]: What `id` is this? Is it the tx nonce?

File: src/mempool.ts [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+            _.keys(this.nodeRegionByUrl),
+            async nodeUrl => {
+                const transactions = await this.getSingleTxPoolContentAsync(nodeUrl);
+                for (const transaction of transactions) {
+                    await this.db.putMempoolTxIfNotExists(transaction.hash, nodeUrl);
+                }
+                return transactions;
+            },
+        );
+        const txPools = (await Promise.all(txPoolPromises));
```
Comment (fabioberger) [inline src/mempool.ts]: Unnecessary outer parens.

File: src/interval_utils.ts [hunk 15]
```diff
@@ -0,0 +1,20 @@
+import * as _ from 'lodash';
```
Comment (fabioberger) [inline src/interval_utils.ts]: Let's nest this file and `utils` inside of a `utils` folder.

File: src/mempool.ts [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+            await this.db.putNodeIfNotExists(nodeUrl, this.nodeRegionByUrl[nodeUrl], nodeVersion);
+        });
+    }
+    public async getTxPoolContentAsync(): Promise<CanonicalMempoolTransaction[]> {
+        const txPoolPromises = _.map(
+            _.keys(this.nodeRegionByUrl),
+            async nodeUrl => {
+                const transactions = await this.getSingleTxPoolContentAsync(nodeUrl);
+                for (const transaction of transactions) {
+                    await this.db.putMempoolTxIfNotExists(transaction.hash, nodeUrl);
```
Comment (fabioberger) [inline src/mempool.ts]: Hm, so perf. wise, we are first inserting all the txs to the DB before returning the list to be cached in memory. As we discussed, I think we should *first* update the in-memory cache and *then* update the DB.

File: src/mempool.ts [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+        const txPools = (await Promise.all(txPoolPromises));
+        const mempoolContent = _.flatten(txPools);
+        const mempoolContentWithoutDuplicates = _.uniqBy(mempoolContent, 'hash');
+        utils.consoleLog('Combined', mempoolContentWithoutDuplicates.length);
+        return mempoolContentWithoutDuplicates;
+    }
+    private async getSingleTxPoolContentAsync(nodeUrl: string): Promise<CanonicalMempoolTransaction[]> {
+        let isSyncing;
+        try {
+            isSyncing = await this.isSyncing(nodeUrl);
```
Comment (fabioberger) [inline src/mempool.ts]: `isSyncingAsync` -> Forgetting `async` suffix in lots of places.

File: src/mempool.ts [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+        } catch (e) {
```
Comment (fabioberger) [inline src/mempool.ts]: Use `err` not `e`

File: src/mempool.ts [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+            utils.consoleLog(this.nodeRegionByUrl[nodeUrl], `Down!`);
```
Comment (fabioberger) [inline src/mempool.ts]: Let's add a TODO comment here to log this to Rollbar or to notify the devs some how.

File: src/mempool.ts [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+            const syncProgress = await this.getSyncProgress(nodeUrl);
+            const percentage = Math.floor(100 * (syncProgress.currentBlock / syncProgress.highestBlock));
+            utils.consoleLog(this.nodeRegionByUrl[nodeUrl],
+                `Syncing: ${percentage}% (${syncProgress.currentBlock}/${syncProgress.highestBlock})`);
+            return [];
+        }
+        const nodeVersion = this.nodeVersion[nodeUrl];
+        let transactions;
+        if (utils.isParityNode(nodeVersion)) {
+            transactions = this.getParityTxPoolContentAsync(nodeUrl);
```
Comment (fabioberger) [inline src/mempool.ts]: Add `await`

File: src/mempool.ts [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+        } else if (utils.isGethNode(nodeVersion)) {
+            transactions = this.getGethTxPoolContentAsync(nodeUrl);
```
Comment (fabioberger) [inline src/mempool.ts]: Add `await`

File: src/mempool.ts [hunk 22]
```diff
@@ -0,0 +1,186 @@
...
+        } else {
+            throw new Error(`Transaction pool watching is only available when \
+                             connected to Geth or Parity node. You're currently connected to: ${nodeVersion}`);
```
Comment (fabioberger) [inline src/mempool.ts]: Nice!

File: src/mempool.ts [hunk 23]
```diff
@@ -0,0 +1,186 @@
...
+        }
+        return transactions;
+    }
+    private async getGethTxPoolContentAsync(nodeUrl: string): Promise<CanonicalMempoolTransaction[]> {
+        const payload = {
+            method: 'txpool_content',
+            params: [],
+            id: this.id++,
```
Comment (fabioberger) [inline src/mempool.ts]: Not a fan of this pattern. Does this return the new increment? Can we do this on it's own line?

File: src/mempool.ts [hunk 24]
```diff
@@ -0,0 +1,186 @@
...
+            jsonrpc: '2.0',
```
Comment (fabioberger) [inline src/mempool.ts]: Put the version string in a constant at top of the file.

File: src/mempool.ts [hunk 25]
```diff
@@ -0,0 +1,186 @@
...
+        };
+        const response = await this.sendRawPayloadAsync(payload, nodeUrl) as GethTxPool;
+        const pending = response.pending;
+        const transactions = [];
+        for (const txsByNonce of _.values(pending)) {
```
Comment (fabioberger) [inline src/mempool.ts]: Put `_.values(pending)` on separate line.

File: src/mempool.ts [hunk 26]
```diff
@@ -0,0 +1,186 @@
...
+        const payload = {
+            method: 'web3_clientVersion',
+            params: [],
+            id: this.id++,
+            jsonrpc: '2.0',
+        };
+        const result = await this.sendRawPayloadAsync(payload, nodeUrl);
+        return result;
+    }
+    private async isSyncing(nodeUrl: string): Promise<boolean> {
```
Comment (fabioberger) [inline src/mempool.ts]: `isSyncingAsync`

File: src/mempool.ts [hunk 27]
```diff
@@ -0,0 +1,186 @@
...
+        const payload = {
+            method: 'eth_syncing',
+            params: [],
+            id: this.id++,
+            jsonrpc: '2.0',
+        };
+        const result = await this.sendRawPayloadAsync(payload, nodeUrl);
+        return result;
+    }
+    private async getSyncProgress(nodeUrl: string): Promise<SyncProgress> {
```
Comment (fabioberger) [inline src/mempool.ts]: `getSyncProgressAsync`

File: src/mempool.ts [hunk 28]
```diff
@@ -0,0 +1,186 @@
...
+              'Accept': 'application/json',
+              'Content-Type': 'application/json',
+            },
+        });
+        const result = (await resp.json()).result;
+        return result;
+    }
+    private parityTransactionToCanonical(tx: ParityMempoolTransaction): CanonicalMempoolTransaction {
+        const canonicalMempoolTransaction: CanonicalMempoolTransaction = {
+            ..._.pick(tx, BASE_MEMPOOL_TRANSACTION_KEYS),
```
Comment (fabioberger) [inline src/mempool.ts]: Assign to var on separate line please.

File: src/mempool.ts [hunk 29]
```diff
@@ -0,0 +1,186 @@
...
+            blockHash: null,
+            transactionIndex: null,
```
Comment (fabioberger) [inline src/mempool.ts]: Why are these here?

File: src/types.ts [hunk 30]
```diff
@@ -0,0 +1,49 @@
...
+    transactionIndex: null;
+    condition: null;
+    creates: null;
+    networkId: number;
+    publicKey: string;
+    raw: string;
+    standardV: string;
+}
+
+// You see what happens Larry!?
```
Comment (fabioberger) [inline src/types.ts]: Ke?

File: test/mempool_test.ts [hunk 31]
```diff
@@ -0,0 +1,197 @@
...
+import * as Sinon from 'sinon';
+import * as firstBy from 'thenby';
+import {chaiSetup} from './utils/chai_setup';
+import {CanonicalMempoolTransaction} from '../src/types';
+
+chaiSetup.configure();
+const expect = chai.expect;
+
+describe('API tests', () => {
+    let Mempool: any;
```
Comment (fabioberger) [inline test/mempool_test.ts]: Why any?

File: test/mempool_test.ts [hunk 32]
```diff
@@ -0,0 +1,197 @@
...
+        });
+        mockedFetch.postOnce(GETH_NODE_URL, {result: IS_SYNCING});
+        mockedFetch.postOnce(GETH_NODE_URL, {result: gethMempool});
+        const nodeRegionByUrl = {
+            [PARITY_NODE_URL]: 'FRA',
+            [GETH_NODE_URL]: 'Tokyo',
+        };
+        const dbMock = {
+            putNodeIfNotExists: async () => undefined,
+            putMempoolTxIfNotExists: async () => undefined,
```
Comment (fabioberger) [inline test/mempool_test.ts]: Why not use `_.noop`

File: migrations/001-initial.sql (hunk 2 above)
Comment (LogvinovLeon) [reply migrations/001-initial.sql]: I assume, that we'll use AWS public DNS or ip address as node_url. Those change when you recreate the machine.

File: package.json (hunk 3 above)
Comment (LogvinovLeon) [reply package.json]: Actually - this is not a library - so dep - devDep separation doesn't make sense at all.
And we only use types from BigNumber so here it is rather a devDep

File: src/index.ts (hunk 7 above)
Comment (LogvinovLeon) [reply src/index.ts]: Yes, cause this module exports a function

File: src/index.ts (hunk 9 above)
Comment (LogvinovLeon) [reply src/index.ts]: Constructors can't we async :'(
We can do `fireAndForget` but I'd prefer this.

File: src/db.ts [hunk 33]
```diff
@@ -5,9 +5,8 @@ export class DB {
     constructor(db: Database) {
         this.db = db;
     }
-    public async getTransactions(): Promise<any> {
-        const transactions = await this.db.all('SELECT * FROM tx');
-        return transactions;
+    public async enableCascadingAsync(): Promise<void> {
```
Comment (fabioberger) [inline src/db.ts]: Let's add a comment of why this is needed.

File: src/db.ts [hunk 34]
```diff
@@ -19,4 +18,14 @@ export class DB {
...
         );
     }
+    public async addTransactionsFromBlockAsync(blockHash: string, transactions: string[]): Promise<void> {
+        const timestamp = Date.now();
+        await this.db.run('INSERT INTO block VALUES (?, ?)', [blockHash, timestamp]);
+        for (const txHash of transactions) {
+            await this.db.run('INSERT INTO tx VALUES (?, ?)', [txHash, blockHash]);
+        }
+    }
+    public async removeTransactionsFromBlockAsync(blockHash: string, transactions: any[]): Promise<void> {
```
Comment (fabioberger) [inline src/db.ts]: Not using the `transactions` param.

File: src/db.ts [hunk 35]
```diff
@@ -19,4 +18,14 @@ export class DB {
...
+        await this.db.run('DELETE FROM block WHERE hash = ?', [blockHash]);
```
Comment (fabioberger) [inline src/db.ts]: I feel like we should query to see if cascading was enabled on the DB and if not, throw an assertion error here. If cascading isn't enabled (and it's not by default when instantiating an DB class instance), then this method won't actually remove the txs.

File: src/rpc.ts [hunk 36]
```diff
@@ -0,0 +1,115 @@
...
+        const payload = {
+            method: 'eth_syncing',
+            params: [],
+            id: this.jsonRpcRequestId++,
+            jsonrpc: '2.0',
+        };
+        const result = await this.sendRawPayloadAsync(payload, nodeUrl);
+        const syncProgress = {
+            currentBlock: parseInt(result.currentBlock, 16),
+            highestBlock: parseInt(result.highestBlock, 16),
```
Comment (fabioberger) [inline src/rpc.ts]: Instead of repeating `16` here, set to var.
